### PR TITLE
Arrange models

### LINKED ISSUE #41: Arranging models to seperate files
#40

### PR DESCRIPTION
**Pull request by RonHachmon**
arranging models into seperate file
arranging test for each model into seperate file and classes

depends on #35 
fix #41

### CODE REVIEW

**Review by MortarDefender — DISMISSED** [review]
Looks Better Organized,
Great Job.

**Review by MortarDefender — APPROVED** [review]

**Review by YaelD — DISMISSED** [review]
Looks good.
I think it's a better design than before

**Comment by MortarDefender** [comment]
@kobihk @brochwerger @nemerna
Please Pay Attention to the comment in the tests.py, models.py and all of the tests in the class_test folder file

As a result of trying to minimize duplicate code, we imported fixtures and other methods that can help in the tests.
Because Flase8 doesn't "understand" that fixtures are a function that can be added to a function as part of her arguments,
it gives a warning like F401 that says the imported function is not used.

So we found a way to ignore the Flake8 wrong warning by inserting the comment "# noqa: ",
if we remove those comments our code will run and pass the tests but will not pass Flake8 tests.

File: events/class_tests/meeting_tests.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+@pytest.mark.django_db
+def test_delete_persist_possible_meeting(persist_possible_meeting):
+    persist_possible_meeting.delete()
+    assert persist_possible_meeting not in PossibleMeeting.objects.all()
+
+
+@pytest.mark.django_db
+def test_delete_participant_deletes_possible_meeting(persist_possible_meeting):
+    persist_possible_meeting.participant_id.delete()
+    assert persist_possible_meeting not in PossibleMeeting.objects.all()
```
Comment (MortarDefender) [inline events/class_tests/meeting_tests.py]: Because of the wide use in @pytest.mark.django_db
I would suggest creating a class for all the tests in this file and placing this decorator above the class.
```suggestion
@pytest.mark.django_db
class MeetingTests():
   pass 
```

File: events/class_tests/event_tests.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+@pytest.mark.django_db
+def test_add_event(persist_event):
+    assert persist_event in Event.objects.all()
+
+
+@pytest.mark.django_db
+def test_delete_event(persist_event):
+    persist_event.delete()
+    assert persist_event not in Event.objects.all()
+
```
Comment (MortarDefender) [inline events/class_tests/event_tests.py]: Because of the wide use in @pytest.mark.django_db
I would suggest creating a class for all the tests in this file and placing this decorator above the class.
```suggestion
@pytest.mark.django_db
class EventTests():
   pass 
```

File: events/class_tests/participant_test.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+@pytest.fixture
+def event_participant_creator(new_event, user0):  # noqa: F811
+    return create_event_participant(new_event, user0, True)
+
+
+@pytest.fixture
+def event_participant_not_creator(new_event, user0):  # noqa: F811
+    return create_event_participant(new_event, user0, False)
+
```
Comment (MortarDefender) [inline events/class_tests/participant_test.py:21]: Because of the wide use in @pytest.mark.django_db
I would suggest creating a class for all the tests in this file and placing this decorator above the class.
```suggestion
@pytest.mark.django_db
class ParticipantTests():
   pass 
```

File: events/class_models/meeting_models.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        return super().clean()
+
+    def save(self, *args, **kwargs):
+        self.clean()
+        return super().save(*args, **kwargs)
+
+
+def validate_unique_possible_meeting(participant_id, date_time_start, date_time_end):
+    if PossibleMeeting.objects.filter(participant_id=participant_id, date_time_start=date_time_start,
+                                      date_time_end=date_time_end):
```
Comment (MortarDefender) [inline events/class_models/meeting_models.py]: In my opinion, this is a more readable format

```suggestion
    if PossibleMeeting.objects.filter(
        participant_id=participant_id, 
        date_time_start=date_time_start,
        date_time_end=date_time_end
    ):
```

File: events/class_models/event_models.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        PINK = "#FFC0CB", "Pink"
+        ORANGE = "#FFA500", "Orange"
+        YELLOW = "#FFFF00", "Yellow"
+        PURPLE = "#800080", "Purple"
+        GREEN = "#008000", "Green"
+        BLUE = "#0000FF", "Blue"
+        BROWN = "#A52A2A", "Brown"
+        WHITE = "#FFFFFF", "White"
+        GRAY = "#808080", "Gray"
+        BLACK = "#000000", "Black"
```
Comment (MortarDefender) [inline events/class_models/event_models.py]: Because we are going to reference this class in the view then I would suggest putting this class outside of the Event class.

File: events/class_models/event_models.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    title = models.CharField(max_length=50)
+    location = models.CharField(blank=True, max_length=50)
+    description = models.TextField(blank=True)
+    date_time_start = models.DateTimeField()
+    date_time_end = models.DateTimeField()
```
Comment (MortarDefender) [inline events/class_models/event_models.py:24]: As @nemerna suggested in our demo,
I think that we can add default values to the title, date_time_start and date_time_end.
The default difference between the start and end of the event can be 30 minutes.

There are two ways you can implement it, 
1. set a default in the fields, (may cause some issues)
2. when validating if the field is left blank you can set the value of now and now + 30 minutes.

**Review by MortarDefender — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 7]
```diff
@@ -13,6 +13,7 @@ db.sqlite3
 
 
 # Distribution / packaging
+.idea
```
Comment (kobihk) [inline .gitignore]: I think this is an example of something small that you can do in different PR and push it to merge easily

File: events/models.py [hunk 8]
```diff
@@ -1,99 +1,3 @@
...
-        raise ValidationError('Ending date cannot be blank')
-    if starting_date >= ending_date:
-        raise ValidationError(f'{time_format(starting_date)} must be smaller than {time_format(ending_date)}')
-
-
-def time_format(date_time):
-    return date_time.strftime("%Y-%m-%d %H:%M:%S")
+from .class_models.event_models import *  # noqa: F403 F401
+from .class_models.participant_model import *  # noqa: F403 F401
+from .class_models.meeting_models import *  # noqa: F403 F401
```
Comment (kobihk) [inline events/models.py:3]: From my POV sometimes it's good and convenient to have a few modules in the same file instead of separating them into a few files.
but if it bothers you so NP from my side

**Review by kobihk — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 7 above)
Comment (MortarDefender) [reply .gitignore]: Done.
A new PR has been created.

File: events/class_models/event_models.py (hunk 6 above)
Comment (RonHachmon) [reply events/class_models/event_models.py:24]: Hi Tamir ,
1. IMHO title should be a a required field since it aint easy to find a collective name that can describe most of the meeting, 
2. i think default dates values and tests should be done under different PR (but dont get me wrong i would more than happy to it 😃 )

File: events/class_tests/meeting_tests.py (hunk 1 above)
Comment (RonHachmon) [reply events/class_tests/meeting_tests.py]: Done

File: events/class_tests/participant_test.py (hunk 3 above)
Comment (RonHachmon) [reply events/class_tests/participant_test.py:21]: Done

File: events/class_models/event_models.py (hunk 5 above)
Comment (RonHachmon) [reply events/class_models/event_models.py]: Done

File: .gitignore (hunk 7 above)
Comment (RonHachmon) [reply .gitignore:16]: Done

File: events/class_models/event_models.py (hunk 6 above)
Comment (MortarDefender) [reply events/class_models/event_models.py:24]: Ok, start working on it. 😁

**Review by MortarDefender — DISMISSED** [review]
Looks better,
Great work 👍

File: events/class_tests/event_tests.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    return Event(title=title, date_time_start=date_time_start, date_time_end=date_time_end)
+
+
+@pytest.mark.django_db
+class TestEvent():
+
+    def test_add_event(self, persist_event):
+        assert persist_event in Event.objects.all()
+
+    @pytest.mark.django_db
```
Comment (RavidYael) [inline events/class_tests/event_tests.py]: I think you can do without this line, as well without line 44 (:

File: events/class_tests/event_tests.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        ('', DATE_TIME_START, DATE_TIME_END, 'Title cannot be blank'),
+        (TITLE, None, DATE_TIME_END, 'Starting date cannot be blank'),
+        (TITLE, DATE_TIME_START, None, 'Ending date cannot be blank'),
+        (TITLE, DATE_TIME_START, DATE_TIME_START,
+         f'{time_format(DATE_TIME_START)} must be smaller than {time_format(DATE_TIME_START)}'),
+        (TITLE, DATE_TIME_END, DATE_TIME_START,
+         f'{time_format(DATE_TIME_END)} must be smaller than {time_format(DATE_TIME_START)}')
+    ])
+    def test_invalidation(self, title, date_time_start, date_time_end, expected_error):
+        current_error = ''
```
Comment (RavidYael) [inline events/class_tests/event_tests.py]: In my opinion it would be more readable using pytest raises (MATCH=..)

*Just as @RonHachmon recommended me do at the EventFileTest

File: events/class_tests/event_tests.py (hunk 9 above)
Comment (RonHachmon) [reply events/class_tests/event_tests.py]: Done :)

File: events/class_tests/event_tests.py (hunk 10 above)
Comment (RonHachmon) [reply events/class_tests/event_tests.py]: Done

**Review by MortarDefender — DISMISSED** [review]

File: events/models.py (hunk 8 above)
Comment (RonHachmon) [reply events/models.py:3]: We think it will be more easy to maintain and read by seperating the model and test into different directorys

File: events/class_models/event_models.py (hunk 6 above)
Comment (RonHachmon) [reply events/class_models/event_models.py:24]: on it :)

**Review by RavidYael — DISMISSED** [review]
Well done :)
Changes approved

File: .gitignore (hunk 7 above)
Comment (kobihk) [inline .gitignore]: please rebase.
this change already merged in:
https://github.com/redhat-beyond/lets-meet/pull/50

**Review by kobihk — CHANGES_REQUESTED** [review]

**Review by MortarDefender — APPROVED** [review]

**Review by kobihk — APPROVED** [review]

**Review by YaelD — APPROVED** [review]
LGTM

File: .gitignore (hunk 7 above)
Comment (RonHachmon) [reply .gitignore]: Done
